src/mono/mono/mini/method-to-ir.c

```diff
@@ -7463,7 +7464,18 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
 			HandleCallData cdata;
 			memset (&cdata, 0, sizeof (HandleCallData));
 
-			cmethod = mini_get_method (cfg, method, token, NULL, generic_context);
+			// The method to be called may have already been resolved when handling a previous opcode. In that
+			// case, we ignore the operand and act as CALL, instead of CALLVIRT.
+			// E.g. #32166 (box+callvirt optimization)
+			if(cmethod_override) {
```

Comment:
I wish we didn't have to use this - effectively global - variable.  But I guess the call logic is so complicated that it wouldn't be reasonable to do the optimization inplace in the `BOX` case of the switch.

Reply:
I'll see how practical it would be to pull the `call`/`callvirt` logic out to a function. At first glance it looks like a function with lots of parameters, but I'll try.

Reply:
I think we tried a few years ago and it regressed JIT perf. but maybe the C compilers got better...

I wouldn't recommend spending a ton of time on this. I was just making an observation